socketio/src/asynchronous/client/builder.rs

```diff
@@ -1,34 +1,44 @@
-use futures_util::{future::BoxFuture, StreamExt};
+use crate::{error::Result, Event, Payload, TransportType};
+use futures_util::future::BoxFuture;
 use log::trace;
 use native_tls::TlsConnector;
 use rust_engineio::{
     asynchronous::ClientBuilder as EngineIoClientBuilder,
     header::{HeaderMap, HeaderValue},
 };
 use std::collections::HashMap;
-use url::Url;
+use std::sync::Arc;
+use tokio::sync::RwLock;
 
-use crate::{error::Result, Error, Event, Payload, TransportType};
+use url::Url;
 
 use super::{
     callback::{Callback, DynAsyncAnyCallback, DynAsyncCallback},
     client::Client,
+    raw_client::RawClient,
 };
 use crate::asynchronous::socket::Socket as InnerSocket;
 
 /// A builder class for a `socket.io` socket. This handles setting up the client and
 /// configuring the callback, the namespace and metadata of the socket. If no
 /// namespace is specified, the default namespace `/` is taken. The `connect` method
 /// acts the `build` method and returns a connected [`Client`].
+#[derive(Clone)]
 pub struct ClientBuilder {
-    address: String,
-    on: HashMap<Event, Callback<DynAsyncCallback>>,
-    on_any: Option<Callback<DynAsyncAnyCallback>>,
+    pub(crate) address: String,
+    on: Arc<RwLock<HashMap<Event, Callback<DynAsyncCallback>>>>,
+    on_any: Arc<RwLock<Option<Callback<DynAsyncAnyCallback>>>>,
     namespace: String,
     tls_config: Option<TlsConnector>,
     opening_headers: Option<HeaderMap>,
     transport_type: TransportType,
     auth: Option<serde_json::Value>,
+    pub(crate) reconnect: bool,
+    pub(crate) reconnect_on_disconnect: bool,
+    // None reconnect attempts represent infinity.
+    pub(crate) max_reconnect_attempts: Option<u8>,
+    pub(crate) reconnect_delay_min: u64,
+    pub(crate) reconnect_delay_max: u64,
 }
```

Comment:
How comes you're moving all logic to the builder? You're inverting Client and Builder, the Builder has all logic and holds a Client, while the Client just has a builder. That doesn't really make sense.